parallelize/parallelize.go

```diff
@@ -0,0 +1,208 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package parallelize
+
+import (
+	"runtime"
+	"sync"
+)
+
+// NewUnboundedRunner creates unbounded goroutines
+func NewUnboundedRunner(workSize int) *Runner {
+	r := getRunner(workSize)
+	r.wg.Add(workSize)
+	go r.consumeMessagesUnbounded()
+	return r
+}
+
+func getRunner(workSize int) *Runner {
+	r := &Runner{
+		resultQueue: make(chan result, workSize),
+		wg:          &sync.WaitGroup{},
+		workQueue:   make(chan Work, workSize),
+	}
+	return r
+}
+
+func (r *Runner) consumeMessagesUnbounded() {
+	func() {
+		for {
+			wrk, ok := <-r.workQueue
+			if !ok {
+				return
+			}
+
+			go func(wrk Work) {
+				defer r.wg.Done()
+				res, err := wrk.Work()
+				r.resultQueue <- result{
+					data: res,
+					err:  err,
+				}
+			}(wrk)
+		}
+	}()
+}
+
+// NewFixedBoundedRunner creates fixed bounded goroutines by factor of available CPU based on default setting.
+// Default setting is to have parallelization as factor of 4x if iobound work which is supposed to be short.
+// If its a long work use NewBoundedRunner instead with optimal config.
+// If cpu bound work parallel go routine will be bound by cpu.
+func NewFixedBoundedRunner(workSize int, ioBound bool) *Runner {
+	parallelCount := runtime.NumCPU()
```

Comment:
NOTE: this should be using `runtime.GOMAXPROCS(-1)` so that it's aware of container CPU quota as set by our https://github.com/uber-go/automaxprocs/ library

Reply:
so you are saying we should do at top of the parallelize.go underscore imports (_ "go.uber.org/automaxprocs"), and then fetch Num of CPU using runtime.GOMAXPROCS(-1) ?

Reply:
makes sense, to not impose the automaxprocs on this lib users.